### ISSUE TITLE
Remove lingering warnings

### ISSUE THREAD

**Issue by AndrewPaxie** [rollbear/trompeloeil#253]
This issue is relative to Trompeloeil v41.

The following lingering warnings may be resolved,
reducing warnings reported in certain configurations, and
allowing enabled warning flags to increase.

## Resolve warnings from early versions of clang compilers

In the following configurations these warnings are reported.

```text
clang++-3.9     Used default libstdc++-v3 (-10).
                Warnings: -Wundefined-func-template, -Wundefined-internal

clang++-3.8     Used libstdc++-v3-7.  Warnings: -Wundefined-internal
clang++-3.7     Used libstdc++-v3-7.  Warnings: -Wundefined-internal
clang++-3.6     Used libstdc++-v3-7.  Warnings: -Wundefined-internal
clang++-3.5     Used libstdc++-v3-7.  Warnings: -Wundefined-internal
```

- "Used" means "used for all language modes the compiler supports".
- The -x suffix on the library name refers to the version of GCC the library
  ships with.

An example of the `-Wundefined-func-template` warning is from
`clang++-3.9/c++14/libstdc++v3-10`:

```text
./include/trompeloeil.hpp:1247:12:
warning: instantiation of function
'trompeloeil::duck_typed_matcher<
  trompeloeil::lambdas::not_equal,
  nullptr_t
>::operator nullptr_t &&<nullptr_t, void, bool>'
required here, but no definition is available
[-Wundefined-func-template]
    return t == nullptr;
           ^
// :::

./include/trompeloeil.hpp:1161:5:
note: forward declaration of template entity is here
    operator V&&() const;
    ^
./include/trompeloeil.hpp:1247:12:
note: add an explicit instantiation declaration to suppress this warning if
'trompeloeil::duck_typed_matcher<
  trompeloeil::lambdas::not_equal,
  nullptr_t>::operator nullptr_t &&<nullptr_t, void, bool>'
is explicitly instantiated in another translation unit
    return t == nullptr;
           ^
```

An example of `-Wundefined-internal` warning is from
`clang++-3.9/c++14/libstdc++v3-10`:

```text
./include/trompeloeil.hpp:1161:5:
warning: function
'trompeloeil::duck_typed_matcher<
  (anonymous namespace)::(lambda at test/compiling_tests_11.cpp:4023:5),
  const char *,
  const char *>::operator nullptr_t &&<nullptr_t, void, bool>'
has internal linkage but is not defined [-Wundefined-internal]
    operator V&&() const;
    ^
./include/trompeloeil.hpp:1247:12: note: used here
    return t == nullptr;
           ^
```

### Suggested fix

A fix for both warnings is to define the following functions:

```cpp
typed_matcher::operator T() const;
duck_typed_matcher::operator V&&() const;
neg_matcher::operator U() const;
```

This mirrors existing code defining `illegal_argument::operator T() const`.

Note: It is not clear why `ptr_deref::operator U() const` can remain
undefined but does not generate either of these warnings.

## Enable `-Wzero-as-null-pointer-constant` in `CMakeLists.txt`

The `-Wzero-as-null-pointer-constant` warning is disabled in `CMakeLists.txt`
for all configurations, but is only triggered in two configurations:

Compiler: Clang
Versions: 10.x and 11.x
C++ Mode: c++20 (= -std=c++2a)
Library: libstdc++v3-10 (= libstdc++v3 shipped with gcc 10.3.0)

The specific warning comes from a defect in `clang++-10` and `clang++-11`,
and when generated contains the following note:

```text
while rewriting comparison as call to 'operator<=>'
```

See:
bugs.llvm.org,
"Bug 44325 - Wzero-as-null-pointer-constant warns when using `operator<=>`,"
created 17 December 2019.
Available: <https://bugs.llvm.org/show_bug.cgi?id=44325>
Accessed: 12 September 2021

### Suggested fix

Avoid the use of `>=` and `<=` in `is_clamped_lambda`.  The original code
may be compiled in configurations other than those generating this warning.

Original code:

```cpp
auto is_clamped_lambda =
  [](auto x, auto min, auto max) -> decltype(x >= min && x <= max)
  {
    return x >= min && x <= max;
  };
```

Modified code (C++11):

```cpp
auto cxx11_is_clamped_lambda =
  [](std::string x, std::string min, std::string max)
    -> decltype(x.compare(min) >= 0 && x.compare(max) <= 0) {
    return x.compare(min) >= 0 && x.compare(max) <= 0;
  };
```

Modified code (>= C++14):

```cpp
auto is_clamped_lambda =
  [](auto x, auto min, auto max)
  -> decltype(x.compare(min) >= 0 && x.compare(max) <= 0)
  {
    return x.compare(min) >= 0 && x.compare(max) <= 0;
  };
```

**Comment by rollbear** [comment]
This one is done, isn't it @AndrewPaxie?

**Comment by AndrewPaxie** [comment]
Yes this was fixed with the merge of #254.

**Comment by rollbear** [comment]
Fix included in release v42
